libs/s25main/network/GameServer.cpp

```diff
@@ -430,13 +433,111 @@ void GameServer::Stop()
+    for(unsigned player = 0; player < playerInfos.size(); ++player)
+    {
+        JoinPlayerInfo& playerInfo = playerInfos[player];
+        switch(playerInfo.team)
+        {
+            case TM_RANDOMTEAM:
+            case TM_TEAM1: ++nPlayers[0]; break;
+            case TM_RANDOMTEAM2:
+            case TM_TEAM2: ++nPlayers[1]; break;
+            case TM_RANDOMTEAM3:
+            case TM_TEAM3: ++nPlayers[2]; break;
+            case TM_RANDOMTEAM4:
+            case TM_TEAM4: ++nPlayers[3]; break;
+            case TM_TEAM_1_TO_2:
+                ++unassignedPlayers;
+                PotentialPlayers[0].insert(player);
+                PotentialPlayers[1].insert(player);
+                break;
+            case TM_TEAM_1_TO_3:
+                ++unassignedPlayers;
+                PotentialPlayers[0].insert(player);
+                PotentialPlayers[1].insert(player);
+                PotentialPlayers[2].insert(player);
+                break;
+            case TM_TEAM_1_TO_4:
+                ++unassignedPlayers;
+                PotentialPlayers[0].insert(player);
+                PotentialPlayers[1].insert(player);
+                PotentialPlayers[2].insert(player);
+                PotentialPlayers[3].insert(player);
+                break;
+            case TM_NOTEAM: break;
+        }
+    }
+    // Check for unassigned players first.
+    if(unassignedPlayers == 0)
+        return false;
+
+    while(unassignedPlayers)
+    {
+        // Set of potential teams for the next player.
+        std::vector<int> teamsForNextPlayer;
+
+        // Determine the minimal team size for teams that can take an unassigned player.
+        unsigned minNextTeamSize = std::numeric_limits<unsigned>::max();
+        for(int team = 0; team < 4; ++team)
+        {
+            // Check if we can add a player to this team at all.
+            if(!PotentialPlayers[team].empty())
+                minNextTeamSize = std::min(minNextTeamSize, nPlayers[team]);
+        }
+
+        for(int team = 0; team < 4; ++team)
+        {
+            // Check if we can add a player to this team at all.
+            if(PotentialPlayers[team].empty())
+                continue;
+            // Check if this is a team with the minimal number of players amongst teams that can have unassigned ones.
+            if(nPlayers[team] > minNextTeamSize)
+                continue;
+            teamsForNextPlayer.push_back(team);
+        }
+
+        RTTR_Assert_Msg(!teamsForNextPlayer.empty(), "Expected to have teams with potential players!");
+        int nextTeam = teamsForNextPlayer[rand() % teamsForNextPlayer.size()];
+        RTTR_Assert_Msg(!PotentialPlayers[nextTeam].empty(), "Expected next team to have potential players!");
+
+        // Pick a random player that can go into this team.
+        auto it = PotentialPlayers[nextTeam].begin();
+        std::advance(it, rand() % PotentialPlayers[nextTeam].size());
+        int nextPlayer = *it;
```

Comment:
I feel this should go into a new function. This is rather verbose and hence hard to see what it does, and used a couple of times. So my suggestion: Create a function `element getRandomElement(collection, prng)` with another overload without the prng (which creates one internally and calls the other function)
Or better: Instead of a prng it could take a functor `getRandIndex(max)` which returns a value in `[0,max]`. We have several other uses in the code where we need a random element from a container, so that would certainly help.

Can be done in a follow-up though.